### PR TITLE
fix(python): converting from numpy datetime64 and overriding dtype with a different resolution was returning incorrect results

### LINKED ISSUE #15991: Datetimes parse incorrectly when converting from pandas with np.datetime64 to polars with a schema that specifies datetime.datetime
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

As of polars 0.20.23 numpy datetime64 parse incorrectly when converting from pandas to polars with a schema that specifies datetime.datetime

```python

import datetime as dt
import numpy as np
import pandas as pd
import polars as pl

d_pd = pd.DataFrame({"t":[np.datetime64("2024-05-01")]})
d_pl = pl.DataFrame(d_pd, schema=[("t", dt.datetime)])

d_pd
d_pl
```

### Log output

```shell
>>> import datetime as dt
>>> import numpy as np
>>> import pandas as pd
>>> import polars as pl
>>>
>>> d_pd = pd.DataFrame({"t":[np.datetime64("2024-05-01")]})
>>> d_pl = pl.DataFrame(d_pd, schema=[("t", dt.datetime)])
>>>
>>> d_pd
           t
0 2024-05-01
>>> d_pl
shape: (1, 1)
┌───────────────────────┐
│ t                     │
│ ---                   │
│ datetime[μs]          │
╞═══════════════════════╡
│ +56301-01-11 00:00:00 │
└───────────────────────┘
>>>
```

### Issue description

Does not occur in polars 0.20.22.

May be related to https://github.com/pola-rs/polars/pull/15933

Specifying schema with pl.Datetime parses correctly.

### Expected behavior

Datetimes match when converting from pandas to polars.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.23
Index type:           UInt32
Platform:             Windows-11-10.0.22621-SP0
Python:               3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               2024.3.1
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.4
nest_asyncio:         <not installed>
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               2.2.2
pyarrow:              16.0.0
pydantic:             2.7.1
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #15991

### CODE REVIEW

File: py-polars/polars/series/series.py [hunk 1]
```diff
@@ -326,10 +326,14 @@ def __init__(
             )
             if values.dtype.type in [np.datetime64, np.timedelta64]:
                 # cast to appropriate dtype, handling NaT values
+                input_dtype = _resolve_temporal_dtype(None, values.dtype)
+                # numpy dtype has already been validated in `numpy_to_pyseries`
+                assert input_dtype is not None
```
Comment (alexander-beedie) [inline py-polars/polars/series/series.py:331]: Don't think we should leave the `assert` here 🤔

Comment (MarcoGorelli) [reply py-polars/polars/series/series.py:331]: I put it for mypy - would you just `# type: ignore` or cast instead?

Comment (alexander-beedie) [reply py-polars/polars/series/series.py:331]: > I put it for mypy - would you just `# type: ignore` or cast instead?

Yup, a `type: ignore[...]` is preferable :sunglasses:👍 
(No runtime cost/behaviour)

**Review by alexander-beedie — APPROVED** [review]

**Comment by MarcoGorelli** [comment]
thanks for your review!

**Pull request merged**